selfdrive/car/mazda/carstate.py

```diff
@@ -0,0 +1,235 @@
+    ("FL", "WHEEL_SPEEDS", 0),
+    ("FR", "WHEEL_SPEEDS", 0),
+    ("RL", "WHEEL_SPEEDS", 0),
+    ("RR", "WHEEL_SPEEDS", 0),
+    ("CRZ_ACTIVE", "CRZ_CTRL", 0),
+    ("STANDSTILL","PEDALS", 0),
+    ("BRAKE_ON","PEDALS", 0),
+    ("GEAR","GEAR", 0),
+    ("DRIVER_SEATBELT", "SEATBELT", 0),
+    ("FL", "DOORS", 0),
+    ("FR", "DOORS", 0),
+    ("BL", "DOORS", 0),
+    ("BR", "DOORS", 0),
+    ("GAS_PEDAL_PRESSED", "CRZ_EVENTS", 0),
+  ]
+
+  checks = [
+    # sig_address, frequency
+    ("BLINK_INFO", 10),
+    ("STEER", 67),
+    ("STEER_RATE", 83),
+    ("STEER_TORQUE", 83),
+    ("WHEEL_SPEEDS", 100),
+    ("CRZ_CTRL", 50),
+    ("CRZ_EVENTS", 50),
+    ("PEDALS", 50),
+    ("SEATBELT", 10),
+    ("DOORS", 10),
+    ("GEAR", 20),
+  ]
+
+  return CANParser(DBC[CP.carFingerprint]['pt'], signals, checks, canbus.powertrain)
+
+def get_cam_can_parser(CP, canbus):
+  # this function generates lists for signal, messages and initial values
+  signals = [
+    # sig_name, sig_address, default
+    #("LINE1",      "CAM_LANETRACK", 0),
+    #("CTR",        "CAM_LANETRACK", -1),
+    #("LINE2",      "CAM_LANETRACK", 0),
+    #("LANE_CURVE", "CAM_LANETRACK", 0),
+    #("SIG1",       "CAM_LANETRACK", 0),
+    #("SIG2",       "CAM_LANETRACK", 0),
+    #("ZERO",       "CAM_LANETRACK", 0),
+    #("SIG3",       "CAM_LANETRACK", 0),
+    #("CHKSUM",     "CAM_LANETRACK", 0),
+
+    ("LKAS_REQUEST",     "CAM_LKAS", 0),
+    ("CTR",              "CAM_LKAS", -1),
+    ("ERR_BIT_1",        "CAM_LKAS", 0),
+    ("LINE_NOT_VISIBLE", "CAM_LKAS", 0),
+    ("BIT_1",            "CAM_LKAS", 0),
+    ("ERR_BIT_2",        "CAM_LKAS", 0),
+    ("BIT_2",            "CAM_LKAS", 0),
+    ("CHKSUM",           "CAM_LKAS", 0),
+
+    ("BIT2",       "CAM_LANEINFO", 1),
+    ("NO_ERR_BIT", "CAM_LANEINFO", 1),
+    ("S1",         "CAM_LANEINFO", 1),
+    ("S1_NOT",     "CAM_LANEINFO", 0),
+  ]
+
+  checks = [
+    # sig_address, frequency
+    ("CAM_LKAS",      16),
+    #("CAM_LANETRACK", 30),
+    ("CAM_LANEINFO", 4),
+  ]
+
+  return CANParser(DBC[CP.carFingerprint]['pt'], signals, checks, canbus.cam)
+
+class CAM_LaneTrack():
+  def __init__(self, ln1, ctr, ln2, lc, s1, s2, z, s3, ck):
+    self.line1 = ln1
+    self.ctr = ctr
+    self.line2 = ln2
+    self.lane_curve = lc
+    self.sig1 = s1
+    self.sig2 = s2
+    self.zero = z
+    self.sig3 = s3
+    self.chksum = ck
+
+class CAM_LaneKAS():
+  def __init__(self, lkas, ctr, er1, lnv, b1, er2, b2, ck):
+    self.lkas = lkas
+    self.ctr = ctr
+    self.err1 = er1
+    self.lnv = lnv
+    self.bit1 = b1
+    self.err2 = er2
+    self.bit2 = b2
+    self.chksum = ck
+
+class STEER_LKAS():
+  def __init__(self):
+    self.block = 1
+    self.track = 1
+    self.handsoff = 0
+
+class CarState():
+  def __init__(self, CP, canbus):
+    # initialize can parser
+    self.CP = CP
+    #self.CAM_LT = CAM_LaneTrack(0, -1, 0, 0, 0, 0, 0, 0, 0)
+    self.CAM_LKAS = CAM_LaneKAS(0, -1, 0, 0, 0, 0, 0, 0)
```

Comment:
the init method isn't really required. Just make sure you set the values the first time you call the car state update by simply setting the ctr to -1.

Reply:
Actually, the classes CAM_LaneTrack, CAM_LaneKAS, STEER_LKAS don't need to be defined.
See how it's handled in selfdrive/car/hyundai for example

Reply:
That was early in the development. I then found out the simpler way, but it was a case of  "if it in't broken, don't fix it" I stopped using most of these msgs, so better to just drop em. will do that.